dali/core/exec/tasking_test.cc

```diff
@@ -0,0 +1,367 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <random>
+#include <vector>
+#include "dali/core/exec/tasking.h"
+
+namespace dali::tasking::test {
+
+inline auto t_now() {
+  return std::chrono::high_resolution_clock::now();
+}
+
+template <typename T>
+inline auto t_plus(T delta) {
+  return std::chrono::high_resolution_clock::now() + std::chrono::duration<T>(delta);
+}
+
+TEST(TaskingTest, ExecutorShutdown) {
+  EXPECT_NO_THROW({
+    Executor ex(4);
+    ex.Start();
+  });
+}
+
+
+TEST(TaskingTest, IndependentTasksAreParallel) {
+  int num_threads = 4;
+  Executor ex(num_threads);
+  ex.Start();
+
+  std::atomic_int parallel;
+  std::atomic_bool done = false;
+  auto timeout = std::chrono::high_resolution_clock::now() + std::chrono::seconds(1);
+  auto complete = Task::Create([](){});
+  for (int i = 0; i < num_threads; i++) {
+    auto task = Task::Create([&]() {
+      if (++parallel)
+        done = true;
+      while (!done && std::chrono::high_resolution_clock::now() < timeout) {}
+      parallel--;
+    });
+    complete->Succeed(task);
+    ex.AddSilentTask(task);
+  }
+  ex.AddSilentTask(complete);
+  ex.Wait(complete);
+  EXPECT_TRUE(done);
```

Comment:
Is there any value in checking if `parallel==0`?

Reply:
I could check it, it's cheap. Also this test is bugged (it should be if `++parallel == num_threads`.